Snap_rounding_2/include/CGAL/Float_snap_rounding_2.h

```diff
@@ -0,0 +1,830 @@
+  Construct_point_at_x_on_segment_2 point_at_x = traits.construct_point_at_x_on_segment_2_object();
+
+  Compare_squared_distance_2 csq_dist_2 = traits.compare_squared_distance_2_object();
+  Squared_round_bound_2 round_bound = traits.squared_round_bound_2_object();
+
+  // Precompute round bounds for points
+  std::vector<double> round_bound_pts;
+  round_bound_pts.reserve(pts.size());
+  for(std::size_t i=0; i<pts.size(); ++i)
+    round_bound_pts.emplace_back(round_bound(pts[i]));
+
+  enum class EVENT_TYPE{INSERT, MIDDLE, REMOVE};
+  struct Event{
+    Event(size_t pi_, size_t li_, EVENT_TYPE t_):pi(pi_),li(li_),type(t_){}
+    size_t pi;
+    size_t li;
+    EVENT_TYPE type;
+  };
+
+  auto event_comparator=[&](Event a, Event b){
+    const Point_2 &pa=pts[a.pi];
+    const Point_2 &pb=pts[b.pi];
+
+    // We sort event lexicographically
+    if(a.pi!=b.pi)
+      return less_xy_2(pa,pb);
+
+    // If two events use the same point, we place remove event first for a smaller y_order range
+    if(a.type != b.type)
+      return a.type > b.type;
+    // Only to have a complete order
+    return a.li < b.li;
+  };
+
+  auto pi_below_li=[&](size_t li, std::pair<size_t, size_t> pair_pi_li){
+    std::size_t ind_pi_support = pair_pi_li.second;
+    std::size_t pi = pair_pi_li.first;
+    if(ind_pi_support==li) return false;
+
+    const Polyline &pl=polylines[li];
+
+    Orientation ori=COLLINEAR;
+    if(pl.front()!=pi && pl.back()!=pi)
+      ori=orientation(pts[pl.front()], pts[pl.back()], pts[pi]);
+    // When collinear, pick an other vertex of the line of pi to decide
+    if(ori==COLLINEAR){
+      const Polyline &pi_support = polylines[ind_pi_support];
+      std::size_t other_point = (pi_support.front()!=pi) ? pi_support.front() : pi_support.back();
+      ori=orientation(pts[pl.front()], pts[pl.back()], pts[other_point]);
+      assert(ori!=COLLINEAR);
+    }
+    return ori==RIGHT_TURN;
+  };
+
+#ifdef DOUBLE_2D_SNAP_FULL_VERBOSE
+  std::cout << "Input points" << std::endl;
+  std::size_t i=0;
+  for(const Point_2 &p: pts)
+    std::cout << i++ << ": " << p << std::endl;
+  std::cout << std::endl;
+  std::cout << "Input polylines" << std::endl;
+  i=0;
+  for(const Polyline &pl: polylines){
+    std::cout << i++ << ":";
+    for(std::size_t pi: pl)
+      std::cout << " " << pi;
+    std::cout << std::endl;
+  }
+  std::cout << std::endl;
+#endif
+
+#ifdef DOUBLE_2D_SNAP_VERBOSE
+  std::cout << "Create the Event Queue" << std::endl;
+#endif
+
+  // Input polylines are supposed going from left to right
+  std::vector<Event> event_queue;
+  event_queue.reserve(2*polylines.size());
+  for(size_t li=0; li<polylines.size(); ++li){
+    Polyline &pl=polylines[li];
+    event_queue.emplace_back(pl.front(), li, EVENT_TYPE::INSERT);
+    for(size_t pi=1; pi<pl.size()-1; ++pi){
+      event_queue.emplace_back(pl[pi], li, EVENT_TYPE::MIDDLE);
+      assert(less_xy_2(pts[pl[pi-1]], pts[pl[pi]]));
```

Comment:
I don't understand how this could hold, except if input are preprocessed to make them "x-monotonic"

Reply:
```
-----           ------
    |     vs    |
    -----   -----
```